### ISSUE TITLE
Add value change options to ControlValueAccessor writeValue and onChange

### ISSUE THREAD

**Issue by JSMike** [angular/angular#20230]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[x] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
A component that implements `ControlValueAccessor` and transforms values to stay in bounds cannot respect `{ emitEvent: false }` option when value patched, it will always emit an event if the `onChange` function is called.

## Expected behavior
If a value is patched to a component that is out-of-bounds and cannot be displayed the component should be able to to transform the incoming value and also respect the option `{ emitEvent: false }`.

[updateControl](https://github.com/angular/angular/blob/1cfa79ca4e21788e0323baf544704ee7ef7d63ea/packages/forms/src/directives/shared.ts#L100) should use an options object like patchValue/setValue.

## Minimal reproduction of the problem with instructions
http://plnkr.co/edit/GyJLgvBkpPdD1y74lZOH?p=preview
In the reproduction you'll notice that when a value outside of the bounds is passed it will update the model. In doing so it will emit a change event to the console even if the new value was patched using `{ emitEvent: false }`.

## What is the motivation / use case for changing the behavior?
The default HTML5 slider will set the value of the slider to the max value if a value higher than the max is set. If I wanted to create a custom slider that followed the same behavior I couldn't do this without emitting an event even if `patchValue(outofboundsvalue, { emitEvent: false });` is used.

## Environment

<pre><code>
Angular version: latest
I'm currently using 4.3.3, example plnkr uses latest.

Browser:
- [x] Chrome (desktop) version 61
This will be the same behavior in all browsers, but I've only tested with Chrome.
</code></pre>

Others:
I'm not sure if this type of writeValue functionality is an anti-pattern or if a custom validator should be used instead. An HTML5 input fields will update a passed value to stay in bounds so I'm assuming Angular form components should have the option to do the same.

Also, not sure if this is a bug or intentional, so I labeled as feature request for now.

**Comment by kara** [comment]
@JSMike `updateControl` is a method that is called as part of the view -> model pipeline. In other words, it should be called when a value change has been initiated by the user from the UI (typing into a text input, for example). In these cases, a valueChange event should always be emitted by design. 

I took a look at your demo, and the reason that your programmatic `patchValue` call is triggering `updateControl` is because of the way your CVA is set up. It's calling the `onChange` method in `writeValue`. I take it this is the "out of bounds" logic you mentioned?  I'd suggest implementing this another way. `writeValue` is the model -> view pipeline, so what you're essentially doing is telling the forms API there has also been a view-> model write every time there is a model -> view write. This isn't recommended. In addition to forcing a valueChange event, you will also force the control to become dirty whenever the value changes programmatically.  Generally speaking, you want to call `onChange` in a CVA *only* when a change has been initiated from the UI (e.g. a mousemove event). So for your out of bounds logic, you would be better off injecting the NgControl into your CVA and programmatically setting its value.

We should really add a more detailed guide on CVAs, but we do have API docs [here](https://angular.io/api/forms/ControlValueAccessor).

**Comment by JSMike** [comment]
> you would be better off injecting the NgControl into your CVA and programmatically setting its value.

@kara even with direct access to the NgControl we would still need a way to forward any options that were used when the control was initially updated.

**Comment by JSMike** [comment]
Hi @kara after speaking at ng-conf you said this issue should be reopened. The flags/options aren't being sent to writeValue in order to pass them forward if a value adjustment is necessary. Also, you mentioned a separate parsers issue that may be able to resolve the value change before reaching the control.

**Comment by kara** [comment]
@JSMike Thanks for taking the time to explain the use case!

**Comment by dario-frongillo** [comment]
Hi,
i have the same issue. My custom component that implement CVA want to fire onchange only when  only when a change has been initiated from the UI and not after a patchvalue. 
In other words i want the same behavior of input htmlt5 using reactive form; if i do a patch value with { emitEvent:  false} on an input html5 no valuechange is fired. Can you please help us to develop a workaround ?
thanks

**Comment by vinagreti** [comment]
Hey @kara, any news or workaround about this issue?
It is a big problem since we cannot use CVA with reactive form.
Thanks a lot!!!

**Comment by Alecell** [comment]
I have this issue. I want to know why when I setValue it just dont send the options as a second parameter on writeValue. I dont get why it dont have this feature. Now I need to make an work around just to do a thing that make so much more sense the framework do.

**Comment by leopangchan** [comment]
Can we please fix it ASAP?? I think it's essential functionality. The patch we have to put in our project causes a nasty technical debt!

**Comment by ud4y** [comment]
@leopangchan I agree. It's not going well with the reactive forms.

**Comment by Yura13** [comment]
Any updates?

**Comment by adaico** [comment]
@JSMike may be I can answer to your question. Emit value or not it is not a ControlValueAccessor responsibility. It is work of FormControl, as you can see from source code method `updateValueAndValidity` ([sourse file](https://github.com/angular/angular/blob/master/packages/forms/src/model.ts)). 

**So main points:** 
- If value is setted from outside code, FormControl emmits it or not based on _emitEvent_ option and sets it through _writeValue_ to ControlValueAccessor view (don't fire _onChange_ in _writeValue_ method - this is the point)
- If value is setted **by user**, you should call _onChange_ callback and it will emmit this event **anyway**.

You can improve your component by rewriting as follows:
`
@Component({
  selector: 'app-slider',
  template: `<input #slider type="range" (change)="setValue($event)">`,
  providers: [SLIDER_VALUE_ACCESSOR]
})
export class SliderComponent implements ControlValueAccessor {
  ...
  setValue(value) {
    this.onChange(value);
  }
  writeValue(val: number) {
    this.mySlider.nativeElement.value = val;
  }
}
`

**Comment by codeedog** [comment]
I wrote a custom control and have been trying to use `reset(model, { emitEvent: false})` with it. It, too, fails to suppress valueChanges events. If it weren't for an offhand comment in this StackOverflow [Q&A](https://stackoverflow.com/q/45241103/314114) I wouldn't have realized my custom control was contributing to the problem.

Also, in an answer to the reference question, a user suggests adding `skip(1)` (skip the first emission from the value change code), to avoid the problem. This solution feels very hacky; in my program a form can be `reset` multiple times (by the user). When its reset to the same object, no events are emitted (somehow the internal code knows to not emit events). Whereas, when the object changes and is reset, a skip(1) is required.

Very messy code logic. I will have to find a work-around.

I'm running Angular 9.1.9. If anyone would like sample custom control code, you can find a that [here](https://stackoverflow.com/a/62563899/314114).

**Comment by codeedog** [comment]
I found a workaround for my issue (use input events). Here's my journey trying to isolate this issue.

**tl/dr:** I was mid-looking for bug in angular code when I found a workaround, so I stopped and documented my progress.

**Workaround:** use a bind to user input event `(input)="trackChange($event)"` instead of subscription to valueChanges

**NOTE:** I'm seeing this problem with a custom control in the following circumstance: FormGroup > FormArray > FormGroup > FormControl(custom). I have another custom form control under the main form (FormGroup > FormControl(custom)) that does not exhibit this problem. So, something about custom form controls beneath an array or array>group triggers the issue.

Because it appears that the emitEvent option is not relayed through the code when using a CVA, I tried to track down the locations where it might get disconnected. While a regular `<input>` didn't have the problem, but my CVA wrapped `<input>` did, I tried to figure out the difference between the two.

I monkey patched `updateValueAndValidity` (UVAV) on the form control in both cases and dumped a stack trace to the console whenever it was called. What I found was that when it was just the `<input>` component, I could see calls that showed the component being added to the form, but no call during `reset(model, { emitEvent: false })`. When I replaced it with my custom control the extra call came later via the Zone event loop; I'm guessing from a macrotask call. The question then, is what part of the code is introducing this extra call.

Also, when monkey patching, I dumped the call arguments to the console including a default argument when none was given. The extra zone invoke occurred with no arguments. There are many places in the code where UVAV is called with arguments, so those can be ruled out. There are a bunch of other locations where its called with no arguments (model.ts & shared.ts), and some locations where its called with an argument that could possibly be undefined.

Before I was able to make any more progress isolating the code that triggered the extraneous call missing event suppression, I realized that I could track events by binding to user input events as follows: `<input (input)="trackChange($event)">`.

Before finding the workaround, I planned to copy the `formControlName` directive code into my project and chip away at it until I isolated the difference that triggered the extra invocation of `updateValueAndValidity()` when using a custom control.

If you've read this far, if you're an angular user and have a CVA, try using input events to track input field changes instead of the subscription, you can still send them through a Subject, if you're using event based programming. If you're an angular developer working on this bug, I hope the above info provides some direction for where to look in the code for this problem.

**Comment by xavadu** [comment]
Any update on this? Is annoying

**Comment by osnersanchez** [comment]
I also have this same problem, when using a control with my CVA and subscribing to the valueChanges I get double event emission when doing a pathValue or a setValue of my control, but when placing emitEvent: false I avoid the duplicate of the event, but I think this should not be the expected behavior. I'm on the lookout for some not-so-patched solution.

**Comment by adaico** [comment]
@osnersanchez provide code of custom control please

**Comment by osnersanchez** [comment]
> @osnersanchez provide code of custom control please

Here you can see my CVA and double event when executing patchValue. Every 5 seconds I call the pathValue in the form. 
In the console you can see the duplicate.

The attention files are these
```
control-value-accessor.component.ts
my-input-with-class.component.ts
control-value-accessor-extend.directive.ts
```

[https://github.com/osnersanchez/demo-control-value-accessor](url)

[https://stackblitz.com/github/osnersanchez/demo-control-value-accessor](url)

**Comment by osnersanchez** [comment]
> > @osnersanchez provide code of custom control please
> 
> Here you can see my CVA and double event when executing patchValue. Every 5 seconds I call the pathValue in the form.
> In the console you can see the duplicate.
> 
> The attention files are these
> 
> ```
> control-value-accessor.component.ts
> my-input-with-class.component.ts
> control-value-accessor-extend.directive.ts
> ```
> 
> [https://github.com/osnersanchez/demo-control-value-accessor](url)
> 
> [https://stackblitz.com/github/osnersanchez/demo-control-value-accessor](url)

My mistake was to have placed the call of the onChange method in the writeValue method, since when calling the pathValue or setValue method of a control, the writeValue method is automatically fired.

I solved it by putting the method call in onTouch, onChange and writeValue in the set of my value

I already uploaded the change to the repository

**Comment by adaico** [comment]
@osnersanchez yes, **my-input.component** works well, but **onChange** firing in **my-input-with-class.component** is excess.
Now everything is ok?

**Comment by born2net** [comment]
same issue and no solution. for now I have to use an ugly side effect to ignore submits manually after I do patchValue.
Yach, I hope this gets fixed, custom components should be able to by pass the ```this.form.valueChanges``` generated by ``` this.form.patchValue(i_values, {emitEvent: false}); ```

**Comment by xavadu** [comment]
Any update on it?

**Comment by adaico** [comment]
@xavadu may be you can show your custom control code? So you don't fire _onChange_ after _whriteValue_ event and something still go wrong?

**Comment by xavadu** [comment]
Yes no problem, this is a custom picker component, when a value is set with `patch` or `set` and the option `emitEvent: false` `valueChanges` observer emit anyway:

```
import {
    Component,
    OnDestroy,
    Input,
    Output,
    OnChanges,
    SimpleChanges,
    EventEmitter,
    ElementRef,
    AfterViewInit,
    ChangeDetectorRef,
    forwardRef
} from '@angular/core';
import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';

import { take, skip } from 'rxjs/operators';
import { UntilDestroy, untilDestroyed } from '@ngneat/until-destroy';

import { Config, PropMap, PickerOption } from './picker.types';
import { PickerService } from './picker.service';

export const PICKER_VALUE_ACCESSOR: any = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(() => PickerComponent),
    multi: true,
};

@UntilDestroy()
@Component({
    selector: 'cx-picker',
    templateUrl: './picker.component.html',
    styleUrls: ['./picker.component.scss'],
    providers: [
        PICKER_VALUE_ACCESSOR,
        PickerService
    ]
})
export class PickerComponent implements OnChanges, AfterViewInit, OnDestroy, ControlValueAccessor {
    @Input() options: PickerOption<any>[];
    @Input() config: Config;
    @Input() placeholder: string;

    @Output('change') change = new EventEmitter<any>();

    selectedOption: any;
    pm: PropMap;
    collapsed: boolean;

    onChange = (option: any) => {};
    onTouched = () => {};
    disabled: boolean = false;

    template; //TODO: Put type

    constructor(
        public pickerService: PickerService,
        private el: ElementRef,
        private ref: ChangeDetectorRef
    ) {
        this.pickerService.config$.pipe(
            untilDestroyed(this)
        ).subscribe(c => {
            this.pm = c.propMap;
            this.template = c.template;
        });

        this.pickerService.collapsed$.pipe(
            skip(2), // Avoid first assigment changes
            take(1),
            untilDestroyed(this)
        ).subscribe(c => {
            this.onTouched();
        });

        this.pickerService.selectedOption$
            .pipe(untilDestroyed(this))
            .subscribe(option => {
                this.change.emit(option);
                this.selectedOption = option;
                this.onChange(option);
                this.pickerService.close();
            });
    }

    ngOnInit() {
    }

    ngOnChanges(changes: SimpleChanges) {
        if (changes.config && changes.config.currentValue) {
            this.pickerService.updateConfig(changes.config.currentValue);
        }

        if (changes.options && changes.options.currentValue && changes.options.currentValue.length > 0) {
            this.pickerService.updateOptions(changes.options.currentValue);
        }
    }

    ngAfterViewInit() {
        this.pickerService.register(this.el, this.config);

        // Necessary for the template ngSwitch
        this.ref.detectChanges();
    }

    ngOnDestroy() {
        this.pickerService.unregister();
    }

    writeValue(optionValue: any): void {
        this.pickerService.updateSelectedOption(optionValue);
    }

    registerOnChange(fn: any): void {
        this.onChange = fn;
    }

    registerOnTouched(fn: any): void {
        this.onTouched = fn;
    }

    setDisabledState(isDisabled: boolean): void {
        this.disabled = isDisabled;
    }
}
```

**Comment by codeedog** [comment]
@adaico I have a briefer version that also fails linked above that I'm repeating [here](https://stackoverflow.com/questions/50159580/for-custom-form-components-is-it-possible-to-use-defaultvalueaccessor-instead-o/62563899#62563899):

    @Component({
      selector: 'inp-with-units',
      template: `
      <div class="input-group">
        <input type="number" ngDefaultControl/>
        <div class="input-group-append" *ngIf="units.length">
             <span class="input-group-text">{{units}}</span>
        </div>
      </div>
    `})
    export class InpNumWithUnits implements ControlValueAccessor {
      @Input('units') units: string = "";
      @ViewChild(DefaultValueAccessor, {static: true}) dva: DefaultValueAccessor;
    
      constructor(@Self() private ngControl: NgControl) {
        this.ngControl.valueAccessor = this;
      }
    
      // Component uses type 'number'. Angular FormControlDirective handles
      // View => Model type conversion (onChange sends number). We wrap the
      // onChange notifier so we, too, can do a type conversion on the value.
      wrap(fn) { return (v) => fn(v === "" || isNaN(v) ? v : +v) }
    
      // *** ControlValueAccessor Methods
      setDisabledState(d: boolean) { this.dva.setDisabledState(d); }
      writeValue(value: any)       { this.dva.writeValue(value); }
      registerOnChange(fn: any)    { this.dva.registerOnChange(this.wrap(fn)); }
      registerOnTouched(fn: any)   { this.dva.registerOnTouched(fn); }
    }

Usage:

    <inp-with-units formControlName="weight" units="lbs"></inp-with-units>

Replaces this:

    <div class="input-group">
      <input type="number" formControlName="weight"/>
      <div class="input-group-append">
           <span class="input-group-text">lbs</span>
      </div>
    </div>

See the link for the GitHub commentary and more details.

**Comment by adaico** [comment]
@xavadu as I've already said: _writeValue_ can change only view and inner value of valueAccessor, but shouldn't cause _onChange_. In your case PickerService immediately emits _selectedOption$_ and then onChange(option).

The easiest solution for you is to add _emitValue_ flag in services' _updateSelectedOption_ and emit _selectedOption$_ if _emitValue_  is true. Otherwise only change value of _this.state.selectedOption_.
So in _writeValue_ it will be  _this.pickerService.updateSelectedOption(optionValue, { emitValue: false })_
After user change (I couldn't find it in your code) it will be  _this.pickerService.updateSelectedOption(optionValue)_

**Comment by adaico** [comment]
@codeedog I've tried to find something suspicious in your component, but failed. Copied it into my project and made a test - everything work just fine.. reset and setValue emit value only without _emitEvent: false_
`setInterval(() => this.c.reset(Math.random(), { emitEvent: false }), 2000);`

May be I didn't understand the problem? 9th angular

**Comment by codeedog** [comment]
@adaico Thank you for testing it! I was using Angular 9.1.9 (according to my text above), so we're in the same system. I believe my other comment is also important, that is, *where you place the custom control*. You have to place it deep.

> NOTE: I'm seeing this problem with a custom control in the following circumstance: ***FormGroup > FormArray > FormGroup > FormControl(custom)***. I have another custom form control under the main form (FormGroup > FormControl(custom)) that does not exhibit this problem. So, something about custom form controls beneath an array or array>group triggers the issue.

I, too, did not see the behavior all of the time. Sometimes it worked correctly if it was sitting in a top level form group (group > control), but when deep (group > array > group > control) it would still get the notification.

How deep did you place the control?

**Comment by adaico** [comment]
@codeedog yes, I've placed it on top level. Because if CustomAccessor works fine on this level, It is absolutely the same with DefaultValueAccessor for angular forms on deeper levels. 
With deeper levels we should restore situation in more detail. On which level you set value and on which subscribe to changes?

**Comment by codeedog** [comment]
At the same level for both setValue and subscribe to changes, I believe. Reiterating, this only occurred when the custom control was deep, not at the top level of the form. That was reliably reproducible.

**Comment by DFG-kjtroll** [comment]
I ran into what I believe is this same issue or closely related to this issue, in a custom CVA picker component, where the single-select version was triggering valueChanged subscribers, even when the FormGroup was first being created with FormBuilder.group(). None of the other FormControls were emitting anything (including multi-select versions of the same component), even without an emitEvent:false option.

I found a rather clean workaround that I believe applies to this. Set a boolean property, default false, on your CVA component. Only call the onChange function in your CVA component if this boolean is true. Set this boolean to true when the component view is interacted with a.k.a. button click.

This way the value still gets properly set when the formGroup is initialized with a value, or set via FormControl.setValue(), before user interaction, but valueChanged subscribers won't fire unless it has actually been set by user interaction.

**Comment by alexandis** [comment]
I have two form custom controls (drop-down list and trivia textbox) which should change each other's value. So they have `valueChange` handlers. The problem is that I can't prevent value change method call when changing the another's control value from within the handler. As result, I even have a parasite third method call at the end. So I have this chain of calls when changing the textbox from UI:
OnDropdownChange -> OnTextBoxChange -> OnDropdownChange

**Comment by alexandis** [comment]
I have two form custom controls: drop-down list and textbox. Each of those should change the value in the another one. The problem is I can't get rid of "another" control valueChanges triggering despite on `emitEvent: false`. As result, the logic does not work properly.

**Comment by adaico** [comment]
@[alexandis](https://github.com/alexandis), 
I can suppose that later you subscribe to _this.addressForm.valueChanges_ and only there get double emit. Am I right?
In this case you also need to add _onlySelf_ to both setValues flag to avoid parent control emit.

**Comment by alexandis** [comment]
> 

It turned out this was a problem of the custom component implementation: using HTML element or even Google Material component works properly (no event firing). So we gave up.

**Comment by adaico** [comment]
@codeedog may be my comment above can be helpful for you. I didn't understand last time, but as you've told about deeper levels, you could add only _emitEvent: false_ on setValue and then subscribe to outer form.

**Comment by adaico** [comment]
@alexandis ok. May be my comments earlier with examples will help you with implementation.

**Comment by codeedog** [comment]
@adaico I appreciate the suggestion, however I'm pretty sure that doesn't work.

When the custom component is added at a deeper level, it doesn't matter where in the hierarchy one subscribes to the emit event. The problem stems from the location of the CVA in the form hierarchy, not the location of the subscriptions. The code responsible for watching changes and originating the event has a problem when the CVA is placed deep down some levels.

I left thorough notes (Aug 28, 2020) and simplified code (Apr 12, 2021) above to help anyone wishing to track down the source of this problem. I urge you to reread those notes, click through the links to examine the stack overflow code and test the isolated code. I spent a considerable amount of time tracking down the source of the problem (2+ weeks?) and trying to isolate the region of code in Angular responsible for this problem including a workaround.

If there is a simple, that would be great, but I'm suspicious that there really is a bug somewhere and no simple way around it.

**Comment by Harpush** [comment]
In our case the solution was using a flag. We had an internal CVA form that emitted changes on valueChanges which write value caused to fire. We added a boolean and filtered the onChange call based on it. In writeValue we set it to true in the first line and false in the last line thus skipping all onChange emissions from writeValue.

**Comment by anatoliykalinskiy** [comment]
My implementation

<app-custom-input [formControl]="formControlText"></app-custom-input>
<p>{{formControlText.valueChanges | async}}</p>

import {
    Component,
    forwardRef,
    Provider,
    Input,
    Output,
    EventEmitter,
    ChangeDetectorRef,
    ChangeDetectionStrategy,
    AfterViewInit,
  } from '@angular/core';
  import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
    
  const INPUT_CONTROL_VALUE_ACCESSOR: Provider = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(() => CustomInputNewComponent),
    multi: true,
  };
  
  @Component({
    selector: 'app-custom-input',
    providers: [INPUT_CONTROL_VALUE_ACCESSOR],
    changeDetection: ChangeDetectionStrategy.OnPush,
    template: `
    <input
    [(ngModel)]="value"
    [disabled]='disabled'
    (blur)='onTouched()'
    (change)='onValueChange($event)'
  >`,
  })
  export class CustomInputNewComponent implements ControlValueAccessor, AfterViewInit {
    private _isValueInit = false;
    /**
     * @see If updateOnChanges = false - changes will be emitted after focus off
     */
    @Input() updateOnChanges = true;
    private _value!: string | number;
    @Input() set value(data: string | number) {
      this._value = data;
  
      if (this._isValueInit) {
        this._onChange(this.value);
      }
    }
    get value(): string | number {
      return this._value;
    }
    @Output() valueChange: EventEmitter<string | number> = new EventEmitter<string | number>();

    disabled = false;
    private _onChange!: (_: any) => void;
    onTouched!: () => void;
    constructor(
      private readonly changeDetect: ChangeDetectorRef,
      ) { }
  
    writeValue(value: string | number): void {
      this.value = value;
  
      this.changeDetect.detectChanges();
    }
    registerOnChange(fn: (_: unknown) => void): void {
      this._onChange = fn;
  
      this.changeDetect.detectChanges();
    }
    registerOnTouched(fn: () => void): void {
      this.onTouched = fn;
  
      this.changeDetect.detectChanges();
    }
  
    setDisabledState(isDisabled: boolean): void {
      this.disabled = isDisabled;
  
      this.changeDetect.detectChanges();
    }
  
    onValueChange(event: any): void {
      this.value = event.target.value;
      this._onChange(this.value);
      this.onTouched();
  
      this.changeDetect.detectChanges();
    }
  
    ngAfterViewInit(): void {
      if (this.updateOnChanges) {
        this._isValueInit = true;
      }
    }
  }

**Comment by adaico** [comment]
@anatoliykalinskiy it is really hard to read your code this way. Please provide it with tabulation. 
First point is that you use `detectChanges` when you don't need it. You should read some about onPush strategy. 
Second.. implementation of what exactly?

**Comment by anatoliykalinskiy** [comment]
> @anatoliykalinskiy it is really hard to read your code this way. Please provide it with tabulation. First point is that you use `detectChanges` when you don't need it. You should read some about onPush strategy. Second.. implementation of what exactly?

Updated. Is it ok?

**Comment by adaico** [comment]
@anatoliykalinskiy sorry, I didn't understand what's the goal. This thread is about how to avoid extra emit of controlValueAccessor when you add `emitEvent: false` flag to controls `setValue` method. CustomValueAccessor gives an opportunity to do everything you want if you need it and use it with forms. The closest example to yours is input with debounce:

```
@Component({
  selector: 'app-custom-input',
  template: `<input [formControl]="control" />`,
  providers: [{
    provide: NG_VALUE_ACCESSOR,
    useExisting: CustomInputComponent,
    multi: true,
  }],
})
export class CustomInputComponent implements OnInit, OnDestroy, ControlValueAccessor {

  protected control = new FormControl<string>('');

  private onChange = (value: string | null) => {};
  private onTouched = () => {};

  private destroy$ = new Subject<void>();

  ngOnInit() {
    this.control.valueChanges
      .pipe(
        tap(() => this.onTouched()),
        debounceTime(300),
        takeUntil(this.destroy$),
      )
      .subscribe((value: string | null) => {
        this.onChange(value)
      });
  }

  writeValue(value: string): void {
    this.control.setValue(value, { emitEvent: false });
  }

  registerOnChange(onChange: (value: string | null) => void): void {
    this.onChange = onChange;
  }

  registerOnTouched(onChange: () => void): void {
    this.onTouched = onChange;
  }

  setDisabledState(isDisabled: boolean): void {
    isDisabled
      ? this.control.disable({ emitEvent: false })
      : this.control.enable({ emitEvent: false });
  }

  ngOnDestroy() {
    this.destroy$.next();
    this.destroy$.complete();
  }

}
```

The main idea is to fire `onChange` on a user event and not to fire with inner logical changes, because `emitEvent: false` prevents only inner form's event, but not yours. 
If you remove `emitEvent: false` in `writeValue` method in my example, you will get two emits after outer `setEvent(someValue)` and one after `setEvent(someValue, { emitEvent: false })`.

**Comment by julian-payne** [comment]
@adaico 
But your implementation still does not respect any options provided to this control when the control is passed via `[formControl]` or `[formControlName]` `@Input` through the template (or any other way).

It just assumes all model -> view changes are `emitEvent: false` flagged.

**Comment by julian-payne** [comment]
Having further investigating this issue, I figured out that the `writeValue` actually represents "model -> view" change.

This is well demonstrated in the `@angular/material` [checkbox component](https://github.com/angular/components/blob/main/src/material/checkbox/checkbox.ts#L299)'s `writeValue` method.

I've changed my code so that all native events trigger the "model -> view" change (onChangeFn) and the `writeValue`. Not calling the onChangeFn in `writeValue` fixes all my issues.

This really ought to be documented more clearly...
